Mathlib/Algebra/Module/Submodule/Pointwise.lean

```diff
@@ -277,4 +289,229 @@ scoped[Pointwise] attribute [instance] Submodule.pointwiseMulActionWithZero
 
 end
 
+section set_acting_on_submodules
+
+variable [AddCommMonoid M] [Module R M]
+
+/--
+Let `s ⊆ R` be a set and `N ≤ M` be a submodule, then `s • N` is the smallest submodule containing
+all `r • n` where `r ∈ s` and `n ∈ N`.
+-/
+protected def pointwiseSetSMul : SMul (Set R) (Submodule R M) where
+  smul s N := sInf { p | ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → r • n ∈ p }
+
+scoped[Pointwise] attribute [instance] Submodule.pointwiseSetSMul
+
+variable (s : Set R) (N : Submodule R M)
+
+lemma mem_set_smul_def (x : M) :
+    x ∈ s • N ↔
+  x ∈ sInf { p : Submodule R M | ∀ ⦃r : R⦄ {n : M}, r ∈ s → n ∈ N → r • n ∈ p } := Iff.rfl
+
+lemma mem_set_smul_of_mem_mem {r : R} {m : M} (mem1 : r ∈ s) (mem2 : m ∈ N) :
+    r • m ∈ s • N := by
+  rw [mem_set_smul_def, mem_sInf]
+  exact fun _ h => h mem1 mem2
+
+lemma set_smul_le (p : Submodule R M)
+    (closed_under_smul : ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → r • n ∈ p) :
+    s • N ≤ p :=
+  sInf_le closed_under_smul
+
+lemma set_smul_le_iff (p : Submodule R M) :
+    s • N ≤ p ↔
+    ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → r • n ∈ p := by
+  fconstructor
+  · intro h r n hr hn
+    exact h <| mem_set_smul_of_mem_mem s N hr hn
+  · apply set_smul_le
+
+lemma set_smul_eq_of_le (p : Submodule R M)
+    (closed_under_smul : ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → r • n ∈ p)
+    (le : p ≤ s • N) :
+    s • N = p :=
+  le_antisymm (set_smul_le s N p closed_under_smul) le
+
+lemma set_smul_mono_right {p q : Submodule R M} (le : p ≤ q) :
+    s • p ≤ s • q :=
+  set_smul_le _ _ _ fun _ _ hr hm => mem_set_smul_of_mem_mem (mem1 := hr)
+    (mem2 := le hm)
+
+lemma set_smul_mono_left {s t : Set R} (le : s ≤ t) :
+    s • N ≤ t • N :=
+  set_smul_le _ _ _ fun _ _ hr hm => mem_set_smul_of_mem_mem (mem1 := le hr)
+    (mem2 := hm)
+
+lemma set_smul_le_of_le_le {s t : Set R} {p q : Submodule R M}
+    (le_set : s ≤ t) (le_submodule : p ≤ q) : s • p ≤ t • q :=
+  le_trans (set_smul_mono_left _ le_set) <| set_smul_mono_right _ le_submodule
+
+variable {s N} in
+@[elab_as_elim]
+lemma set_smul_inductionOn {prop : M → Prop} (x : M)
+    (hx : x ∈ s • N)
+    (smul₀ : ∀ ⦃r : R⦄ ⦃n : M⦄, r ∈ s → n ∈ N → prop (r • n))
+    (smul₁ : ∀ (r : R) ⦃m : M⦄, prop m → prop (r • m))
+    (add : ∀ ⦃m₁ m₂ : M⦄, prop m₁ → prop m₂ → prop (m₁ + m₂))
+    (zero : prop 0) :
+    prop x :=
+  set_smul_le s N
+    { carrier := {m : M | prop m}
+      add_mem' := fun {x y} ↦ @add x y
+      zero_mem' := zero
+      smul_mem' := smul₁ }
+    smul₀ hx
```

Comment:
Can you generalize this to `{prop : (x : M) → x ∈ s • N → Prop` so that it works with `induction using`?

Reply:
Done in[28d6404](https://github.com/leanprover-community/mathlib4/pull/7140/commits/28d64046205bb588e7dc738321c3bb835505708d)